src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs

```diff
@@ -83,58 +85,63 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
 
             var startupTypes = new List<Type>();
 
-            foreach (var item in extensionItems)
-            {
-                string startupExtensionName = item.Name ?? item.TypeName;
-                _logger.ScriptStartUpLoadingStartUpExtension(startupExtensionName);
+            var functionBindings = _functionMetadataProvider.GetFunctionMetadata(true).SelectMany(f => f.Bindings.Select(b => b.Type));
 
-                // load the Type for each startup extension into the function assembly load context
-                Type extensionType = Type.GetType(item.TypeName,
-                    assemblyName =>
-                    {
-                        if (_builtinExtensionAssemblies.Contains(assemblyName.Name, StringComparer.OrdinalIgnoreCase))
-                        {
-                            _logger.ScriptStartUpBelongExtension(item.TypeName);
-                            return null;
-                        }
-
-                        string path = item.HintPath;
-                        if (string.IsNullOrEmpty(path))
-                        {
-                            path = assemblyName.Name + ".dll";
-                        }
+            foreach (var extensionItem in extensionItems)
+            {
+                if (extensionItem.Bindings.Count == 0 || extensionItem.Bindings.Intersect(functionBindings, StringComparer.OrdinalIgnoreCase).Count() > 0)
```

Comment:
This logic can only be applied when we're dealing with bundles, as this would be a breaking change for all non-bundle customers.